packages/apollo-language-server/src/utilities/graphql.ts

```diff
@@ -87,31 +93,180 @@ export function removeDirectives(ast: ASTNode, directiveNames: string[]) {
+          anyFragmentsRemoved = true;
+          return null;
+        }
+
+        return undefined;
+      }
+    }
+  });
+
+  if (anyFragmentsRemoved) {
+    // We've removed fragments and might have orphaned more fragments, so recursively try to remove more
+    // orphaned fragments.
+    return removeOrphanedFragmentDefinitions(
+      ast,
+      fragmentNamesEligibleForRemoval
+    );
+  }
+
+  return ast;
+}
+
+/**
+ * Recursively remove nodes that have zero-length selection sets
+ */
+function removeNodesWithEmptySelectionSets<AST extends ASTNode>(ast: AST): AST {
+  /**
+   * Flag to know if we need to make another recursive pass
+   */
+  let anyNodesRemoved = false;
+
+  ast = visit(ast, {
+    enter(node) {
+      // If this node _has_ a `selectionSet` and it's zero-length, then remove it.
+      if (
+        "selectionSet" in node &&
+        node.selectionSet != null &&
+        node.selectionSet.selections.length === 0
+      ) {
+        anyNodesRemoved = true;
+        return null;
+      }
+
+      return undefined;
+    }
+  });
+
+  if (anyNodesRemoved) {
+    return removeNodesWithEmptySelectionSets(ast);
+  }
+
+  return ast;
+}
+
+/**
+ * Remove nodes from `ast` when they have a directive in `directiveNames`
+ * @param ast
+ * @param directiveNames
+ */
+export function removeDirectiveAnnotatedFields<AST extends ASTNode>(
+  ast: AST,
   directiveNames: string[]
-) {
+): AST {
+  print;
   if (!directiveNames.length) return ast;
-  return visit(ast, {
-    Field(node: FieldNode): FieldNode | null {
+
+  /**
+   * All fragment definition names we've removed due to a matching directive
+   *
+   * We keep track of these so we can remove associated spreads
+   */
+  const removedFragmentDefinitionNames = new Set<string>();
+
+  /**
+   * All fragment spreads that have been removed
+   *
+   * We can only remove fragment definitions for fragment spreads that we've removed
+   */
+  const removedFragmentSpreadNames = new Set<string>();
+
+  // Remove all nodes with a matching directive in `directiveNames`. Also, remove any operations that now have
+  // no selection set
+  ast = visit(ast, {
+    enter(node) {
+      // Strip all nodes that contain a directive we wish to remove
       if (
+        "directives" in node &&
         node.directives &&
-        node.directives.find(
-          directive =>
-            !!directiveNames.find(name => name === directive.name.value)
+        node.directives.find(directive =>
+          directiveNames.includes(directive.name.value)
         )
-      )
+      ) {
+        // If we're removing a fragment definition then save the name so we can remove anywhere this fragment
+        // was spread
+        if (node.kind === Kind.FRAGMENT_DEFINITION) {
+          removedFragmentDefinitionNames.add(node.name.value);
+        }
+
+        // All nested fragment spreads inside of this definition are now eligible to be removed
+        visit(ast, {
```

Comment:
I don't think we should nest visitors here. This implementation doesn't do what I think you want it to anyway. By revisiting the root ast, instead of the specific node that you are in currently, you aren't "filtering" fragment spreads inside of this definition.

Since the visitor works depth first, you should be able to visit on `FragmentDefinition` to create your list, then visit the FragmentSpread next as long as you sort the document prior to doing the work so that fragment definitions always come first

Reply:
Absolutely right.

Reply:
This appears to still be nested? Why was this resolved?

Reply:
I'm sorry about that; I have no idea. I'll figure this last piece out on Monday! Thanks again @jbaxleyiii !

Reply:
My original logic was to visit all the descendents of `node` to find all fragment spreads being removed because this field was being removed as well. I wrote this to visit `ast` by accident; great catch @jbaxleyiii .

I think that I do, however, want to use a nested `visitor` here to flag all the fragment spreads under this node as eligible for deletion. I've added a comment to hopefully explain this. Please let me know if this is the correct way to handle this problem.